src/ring_theory/polynomial/cyclotomic/basic.lean

```diff
@@ -377,69 +351,75 @@ end
-    rw polynomial.map_prod (int.cast_ring_hom ℂ) (λ i, cyclotomic i ℤ),
-    simp only [int_cyclotomic_spec, polynomial.map_pow, nat.cast_id, map_X, polynomial.map_one,
-               polynomial.map_sub],
-    exact prod_cyclotomic'_eq_X_pow_sub_one hpos
-          (complex.is_primitive_root_exp n (ne_of_lt hpos).symm) },
-  have coerc : X ^ n - 1 = map (int.cast_ring_hom R) (X ^ n - 1),
-  { simp only [polynomial.map_pow, polynomial.map_X, polynomial.map_one, polynomial.map_sub] },
-  have h : ∀ i ∈ n.divisors, cyclotomic i R = map (int.cast_ring_hom R) (cyclotomic i ℤ),
-  { intros i hi,
-    exact (map_cyclotomic_int i R).symm },
-  rw [finset.prod_congr (refl n.divisors) h, coerc,
-      ← polynomial.map_prod (int.cast_ring_hom R) (λ i, cyclotomic i ℤ), integer]
+    simp only [polynomial.map_prod, int_cyclotomic_spec, polynomial.map_pow, map_X,
+               polynomial.map_one, polynomial.map_sub],
+    exact prod_cyclotomic'_eq_X_pow_sub_one hpos (complex.is_primitive_root_exp n hpos.ne') },
+  simpa only [polynomial.map_prod, map_cyclotomic_int, polynomial.map_sub, polynomial.map_one,
+    polynomial.map_pow, polynomial.map_X] using congr_arg (map (int.cast_ring_hom R)) integer
 end
 
-lemma cyclotomic.dvd_X_pow_sub_one (n : ℕ) (R : Type*) [comm_ring R] :
+lemma cyclotomic.dvd_X_pow_sub_one (n : ℕ) (R : Type*) [ring R] :
   (cyclotomic n R) ∣ X ^ n - 1 :=
 begin
+  suffices : cyclotomic n ℤ ∣ X ^ n - 1,
+  { simpa only [map_cyclotomic_int, polynomial.map_sub, polynomial.map_one, polynomial.map_pow,
+      polynomial.map_X] using map_dvd (int.cast_ring_hom R) this },
   rcases n.eq_zero_or_pos with rfl | hn,
   { simp },
-  refine ⟨∏ i in n.proper_divisors, cyclotomic i R, _⟩,
-  rw [←prod_cyclotomic_eq_X_pow_sub_one hn,
-      nat.divisors_eq_proper_divisors_insert_self_of_pos hn, finset.prod_insert],
-  exact nat.proper_divisors.not_self_mem
+  rw [← prod_cyclotomic_eq_X_pow_sub_one hn],
+  exact finset.dvd_prod_of_mem _ (n.mem_divisors_self hn.ne')
 end
 
-open_locale big_operators
-open finset
+lemma prod_cyclotomic_eq_geom_sum {n : ℕ} (h : 0 < n) (R) [comm_ring R] :
+  ∏ i in n.divisors.erase 1, cyclotomic i R = ∑ i in finset.range n, X ^ i :=
+suffices ∏ i in n.divisors.erase 1, cyclotomic i ℤ = ∑ i in finset.range n, X ^ i,
+by simpa only [polynomial.map_prod, map_cyclotomic_int, polynomial.map_sum, polynomial.map_pow,
+  polynomial.map_X] using congr_arg (map (int.cast_ring_hom R)) this,
+by rw [← mul_left_inj' (cyclotomic_ne_zero 1 ℤ), prod_erase_mul _ _ (nat.one_mem_divisors.2 h.ne'),
+  cyclotomic_one, geom_sum_mul, prod_cyclotomic_eq_X_pow_sub_one h]
 
-lemma prod_cyclotomic_eq_geom_sum {n : ℕ} (h : 0 < n) (R) [comm_ring R] [is_domain R] :
-  ∏ i in n.divisors \ {1}, cyclotomic i R = ∑ i in finset.range n, X ^ i :=
+/-- If `p` is prime, then `cyclotomic p R = ∑ i in range p, X ^ i`. -/
+lemma cyclotomic_eq_geom_sum {R : Type*} [ring R] (p : ℕ) [hp : fact p.prime] :
```

Comment:
`R` should probably be explicit here, and I think there is already a `R [ring R]` around. If it's not already declared it is a good idea to do so, it's used in a lot of declarations.

Reply:
BTW, is it more useful to have `cyclotomic (R : Type*) [ring R]` or to have `cyclotomic : int[X]` and use `aeval` to evaluate it? I don't know how often these polynomials are used as polynomials, not as their `eval` functions.

Reply:
Made `R` explicit, also renamed to `cyclotomic_prime`.

Reply:
> BTW, is it more useful to have `cyclotomic (R : Type*) [ring R]` or to have `cyclotomic : int[X]` and use `aeval` to evaluate it? I don't know how often these polynomials are used as polynomials, not as their `eval` functions.

I think it's OK as it is, the theory is mostly algebraic, so it is relevant if `cyclotomic R` is irreducible or stuff like that, I guess it's less common to evaluate it (even if of course sometimes it happens).